File: extra samples/vb/SendToSTU/Form1.vb
'******************************************************* 
'
'  Form1.cs
'  
'  Allow user to select an image and send it to the STU
'
'  Copyright (c) 2015 Wacom GmbH. All rights reserved.
'  
'*******************************************************
Public Class Form1

    Dim tablet As wgssSTU.Tablet
    Dim capability As wgssSTU.ICapability
    Dim capability2 As wgssSTU.ICapability2
    Dim information As wgssSTU.IInformation

    Private Sub btnSendToSTU_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnSendToSTU.Click
        print("SendToSTU")
        SendToSTU()
    End Sub
    Private Function Connect()
        Dim connected = False
        Try
            Dim usbDevices = New wgssSTU.UsbDevices()
            If (usbDevices.Count = 0) Then
                print("No devices found")
            Else
                connected = True
                tablet = New wgssSTU.Tablet()
                Dim ec = tablet.usbConnect(usbDevices(0), True)
                If (ec.value = 0) Then
                    capability = tablet.getCapability()
                    capability2 = tablet.getCapability_2()
                    information = tablet.getInformation()
                    print("Tablet: " + information.modelName)
                Else
                    Throw New Exception(ec.message)
                End If
            End If
        Catch ex As Exception
            print("Exception: " + ex.Message)
        End Try
        Return connected
    End Function
    Private Sub Disconnect()
        Try
            If Not (tablet Is Nothing) Then
                tablet.disconnect()
            End If
        Catch ex As Exception
            print("Exception: " + ex.Message)
        End Try
    End Sub
    Private Sub SendToSTU()
        Try
            If Connect() <> True Then
                Return
            End If

            print("Connected: " + information.modelName)
            Dim protocolHelper = New wgssSTU.ProtocolHelper()
            Dim encodingFlag As wgssSTU.encodingFlag = 0
            Dim encodingMode As wgssSTU.encodingMode = 0

            Dim idP = tablet.getProductId()

            encodingFlag = capability2.encodingFlag
            encodingFlag = protocolHelper.simulateEncodingFlag(idP, encodingFlag)

            print("Encoding flag: " + encodingFlag.ToString())
            If (encodingFlag And wgssSTU.encodingFlag.EncodingFlag_24bit) Then
                If (tablet.supportsWrite()) Then
                    encodingMode = wgssSTU.encodingMode.EncodingMode_24bit_Bulk
                Else
                    encodingMode = wgssSTU.encodingMode.EncodingMode_24bit
                End If

            ElseIf (encodingFlag And wgssSTU.encodingFlag.EncodingFlag_16bit) Then
                If (tablet.supportsWrite()) Then
                    encodingMode = wgssSTU.encodingMode.EncodingMode_16bit_Bulk
                Else
                    encodingMode = wgssSTU.encodingMode.EncodingMode_16bit
                End If
            Else
                ' assumes 1bit is available
                encodingMode = wgssSTU.encodingMode.EncodingMode_1bit
            End If
            print("encodingMode: " + encodingMode.ToString())

            Dim bitmapData  '// This is the flattened data of the bitmap that we send to the device.
            Dim stream As New System.IO.MemoryStream()
            PictureBox1.Image.Save(stream, System.Drawing.Imaging.ImageFormat.Png)

            bitmapData = protocolHelper.resizeAndFlatten(stream.ToArray(),
                            0, 0, 0, 0, capability.screenWidth, capability.screenHeight, CByte(encodingMode), wgssSTU.Scale.Scale_Fit, False, 0)
            tablet.writeImage(encodingMode, bitmapData)

            Disconnect()
        Catch ex As Exception
            print("Exception: " + ex.Message)
        End Try

    End Sub

    Private Sub print(ByVal txt)
        txtDisplay.Text += txt + vbCrLf
        txtDisplay.SelectionStart = txtDisplay.Text.Length ' scroll to end
        txtDisplay.ScrollToCaret()
    End Sub

    Private Sub btnClear_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnClear.Click
        Try
            If Connect() <> True Then
                Return
            End If
            tablet.setClearScreen()
            print("Cleared " + information.modelName)

            Disconnect()
        Catch ex As Exception
            print("Exception: " + ex.Message)
        End Try

    End Sub
    Private Sub btnBrowse_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnBrowse.Click
        Dim res
        res = OpenFileDialog1.ShowDialog()
        If (res = DialogResult.OK) Then
            txtFilename.Text = OpenFileDialog1.FileName
            DisplayImage(Image.FromFile(OpenFileDialog1.FileName))
        End If

    End Sub
    Private Sub DisplayImage(ByVal img As Image)
        ' resize the image to fit the panel
        ' STU displays: 300:396x100 430:320x200 500:640x480 520:800x480 530:800x480
        ' 300/420 scale by 2, else scale by 4. Also handle unexpected size.
        Dim scale = 1
        If (img.Width > 400) Then
            scale = 4
        ElseIf (img.Width > Panel1.Width) Then
            scale = 2
        End If

        PictureBox1.Size = New Size(img.Width / scale, img.Height / scale)
        ' don't exceed the panel size:
        If (PictureBox1.Size.Width > Panel1.Size.Width Or PictureBox1.Size.Height > Panel1.Size.Height) Then
            PictureBox1.Size = Panel1.Size
        End If

        ' centre the image in the panel
        Dim x, y
        x = Panel1.Location.X + ((Panel1.Width - PictureBox1.Width) / 2)
        y = Panel1.Location.Y + ((Panel1.Height - PictureBox1.Height) / 2)
        PictureBox1.Location = New Point(x, y)

        PictureBox1.SizeMode = PictureBoxSizeMode.StretchImage
        PictureBox1.Image = img
        PictureBox1.Parent = Me
        PictureBox1.BringToFront()

    End Sub
    Private Sub Form1_Load(ByVal sender As Object, ByVal e As System.EventArgs) Handles MyBase.Load
        DisplayImage(My.Resources._640x480) ' use resource to simplify and avoid locating a folder
        print("Ready")
    End Sub

    Private Sub btnClose_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnClose.Click
        Me.Close()
    End Sub

End Class

File: extra samples/csharp/SendToSTU/Form1.cs
/******************************************************* 

  Form1.cs
  
  Allow user to select an image and send it to the STU
  Compatible with STU-300, STU-430, STU-500, STU-530 and STU-540
  but not the obsolete STU-541

  Copyright (c) 2015 Wacom GmbH. All rights reserved.
  
********************************************************/
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;

namespace SendToSTU
{
    public partial class Form1 : Form
    {
        private wgssSTU.Tablet tablet;
        private wgssSTU.ICapability capability;
        private wgssSTU.ICapability2 capability2;
        private wgssSTU.IInformation information;


        public Form1()
        {
            InitializeComponent();
            DisplayImage( Properties.Resources._640x480 );  // use resource to simplify and avoid locating a folder
            print("Ready");
        }

        private void btnSendToSTU_Click(object sender, EventArgs e)
        {
            print("Send image to STU display");
            SendToSTU();
        }
        private void print(string txt)
        {
            txtDisplay.Text += txt + "\r\n";
            txtDisplay.SelectionStart = txtDisplay.Text.Length; // scroll to end
            txtDisplay.ScrollToCaret();

        }
        private bool Connect()
        {
            bool connected = false;
            try
            {
                wgssSTU.UsbDevices usbDevices = new wgssSTU.UsbDevices();
                if (usbDevices.Count == 0)
                {
                    print("No devices found");
                }
                else
                {
                    connected = true;
                    tablet = new wgssSTU.Tablet();
                    wgssSTU.IErrorCode ec = tablet.usbConnect(usbDevices[0], true);
                    if (ec.value == 0)
                    {
                        capability = tablet.getCapability();
                        capability2 = tablet.getCapability_2();
                        information = tablet.getInformation();
                        print("Tablet: " + information.modelName);
                    }
                    else
                    {
                        throw new Exception(ec.message);
                    }
                }
            }
            catch (Exception ex)
            {
                print("Exception: " + ex.Message);
            }
            return connected;
        }
        private void Disconnect()
        {
            try
            {
                if (tablet != null)
                    tablet.disconnect();
            }
            catch (Exception ex)
            {
                print("Exception: " + ex.Message);
            }
        }
        private void SendToSTU() {
            try
            {
                if (!Connect())
                {
                    return;
                }
                print("Connected: " + information.modelName);
                wgssSTU.ProtocolHelper protocolHelper = new wgssSTU.ProtocolHelper();
                wgssSTU.encodingFlag encodingFlag = 0;
                wgssSTU.encodingMode encodingMode = 0;

                ushort idP = tablet.getProductId();

                encodingFlag = (wgssSTU.encodingFlag)capability2.encodingFlag;
                encodingFlag = (wgssSTU.encodingFlag)protocolHelper.simulateEncodingFlag(idP, (byte)encodingFlag);

                print("Encoding flag: " + encodingFlag);

                if ((encodingFlag & wgssSTU.encodingFlag.EncodingFlag_24bit) != 0)
                {
                    encodingMode = tablet.supportsWrite() ? wgssSTU.encodingMode.EncodingMode_24bit_Bulk : wgssSTU.encodingMode.EncodingMode_24bit;
                }
                else if ((encodingFlag & wgssSTU.encodingFlag.EncodingFlag_16bit) != 0)
                {
                    encodingMode = tablet.supportsWrite() ? wgssSTU.encodingMode.EncodingMode_16bit_Bulk : wgssSTU.encodingMode.EncodingMode_16bit;
                }
                else
                {
                    // assumes 1bit is available
                    encodingMode = wgssSTU.encodingMode.EncodingMode_1bit;
                }
                print("encodingMode: " + encodingMode);

                byte[] m_bitmapData; // This is the flattened data of the bitmap that we send to the device.
                System.IO.MemoryStream stream = new System.IO.MemoryStream();
                pictureBox1.Image.Save(stream, System.Drawing.Imaging.ImageFormat.Png);
                m_bitmapData = (byte[])protocolHelper.resizeAndFlatten(
                    stream.ToArray(),
                    0, 0, 0, 0,
                    capability.screenWidth, capability.screenHeight,
                    (byte)encodingMode,
                    wgssSTU.Scale.Scale_Fit, 0, 0);
                tablet.writeImage((byte)encodingMode, m_bitmapData);	  // uses the colour mode flags in encodingMode



                Disconnect();
            }
            catch (Exception ex)
            {
                print("Exception: " + ex.Message);
            }

        }

        private void btnClear_Click(object sender, EventArgs e)
        {
            try
            {
                if (!Connect())
                {
                    return;
                }
                tablet.setClearScreen();
                print("Cleared " + information.modelName);
                Disconnect();
            }
            catch (Exception ex)
            {
                print("Exception: " + ex.Message);
            }

        }

        private void btnBrowse_Click(object sender, EventArgs e)
        {
            DialogResult res = openFileDialog1.ShowDialog();
            if (res == DialogResult.OK)
            {
                txtFilename.Text = openFileDialog1.FileName;
                DisplayImage(Image.FromFile(openFileDialog1.FileName));
            }
        }
        private void btnClose_Click(object sender, EventArgs e)
        {
            this.Close();
        }
        
        private void DisplayImage(Image img)
        {
        // resize the image to fit the panel
        // STU displays: 300:396x100 430:320x200 500:640x480 520:800x480 530:800x480
        // 300/420 scale by 2, else scale by 4. Also handle unexpected size.
        int scale = 1;       
        if (img.Width > 400)
            scale = 4;
        else if (img.Width > panel1.Width)
            scale = 2;

        pictureBox1.Size = new Size(img.Width / scale, img.Height / scale);
        // don't exceed the panel size:
        if (pictureBox1.Size.Width > panel1.Size.Width || pictureBox1.Size.Height > panel1.Size.Height)
                    pictureBox1.Size = panel1.Size;     

        //centre the image in the panel
        int x, y;
        x = panel1.Location.X + ((panel1.Width - pictureBox1.Width) / 2);
        y = panel1.Location.Y + ((panel1.Height - pictureBox1.Height) / 2);
        pictureBox1.Location = new Point(x, y);

        pictureBox1.SizeMode = PictureBoxSizeMode.StretchImage;
        pictureBox1.Image = img;
        pictureBox1.Parent = this;
        pictureBox1.BringToFront();
        }
              
    }
}
